### PR TITLE
global_norm  handles empty list and optimizer check

### PR DESCRIPTION
**Pull request by hnyu**
Another unsolved issue is that get_optimizer_info won't obtain optimizer info for a subalgorithm that's an ignored attribute (e.g., self-training), because setup_optimizers will not include ignored attributes. So TB optimizers tab won't show anything for it .

### CODE REVIEW

File: alf/utils/tensor_utils.py [hunk 1]
```diff
@@ -142,6 +142,8 @@ def global_norm(tensors):
     """
     assert alf.nest.is_nested(tensors), "tensors must be a nest!"
     tensors = alf.nest.flatten(tensors)
+    if not tensors:
+        return torch.zeros((), dtype=torch.float32)
```
Comment (hnyu) [inline alf/utils/tensor_utils.py:146]: If not check here, torch.sqrt will complain the arg is int for an empty list.

File: alf/trainers/policy_trainer.py [hunk 2]
```diff
@@ -101,8 +101,9 @@ def __init__(self, config: TrainerConfig):
         self._random_seed = config.random_seed
         self._num_iterations = config.num_iterations
         self._num_env_steps = config.num_env_steps
-        assert self._num_iterations + self._num_env_steps > 0, \
-            "Must provide #iterations or #env_steps for training!"
+        assert (self._num_iterations + self._num_env_steps > 0
+                and self._num_iterations * self._num_env_steps == 0), \
```
Comment (hnyu) [inline alf/trainers/policy_trainer.py:105]: These two need to be exclusive because otherwise the checkpoint interval will function incorrectly.

File: alf/utils/tensor_utils.py (hunk 1 above)
Comment (emailweixu) [reply alf/utils/tensor_utils.py:146]: why not return 0 here?

File: alf/algorithms/sac_algorithm_test.py [hunk 3]
```diff
@@ -223,7 +223,6 @@ def test_sac_algorithm_discrete(self, use_parallel_network):
             use_parallel_network=use_parallel_network,
             env=env,
             config=config,
-            actor_optimizer=alf.optimizers.Adam(lr=1e-2),
```
Comment (emailweixu) [inline alf/algorithms/sac_algorithm_test.py:226]: why is this removed?

File: alf/utils/tensor_utils.py (hunk 1 above)
Comment (hnyu) [reply alf/utils/tensor_utils.py:146]: This function can potentially be used for other purposes so it's better to return a tensor.

File: alf/algorithms/sac_algorithm_test.py (hunk 3 above)
Comment (hnyu) [reply alf/algorithms/sac_algorithm_test.py:226]: In the discrete case, there is no actor network. So this optimizer will not be used. Now any unused optimizer will be checked and produce an error.

File: alf/algorithms/algorithm.py [hunk 4]
```diff
@@ -1002,8 +1007,14 @@ def update_with_gradient(self, loss_info, valid_masks=None, weight=1.0):
 
         all_params = []
         for optimizer in optimizers:
+            params = []
             for param_group in optimizer.param_groups:
-                all_params.extend(param_group['params'])
+                params.extend(param_group['params'])
+            assert params, (
+                "The recorded optimizer '" + optimizer.name +
+                "' haven't been used for learning any parameters! Please check."
```
Comment (Haichao-Zhang) [inline alf/algorithms/algorithm.py:1015]: haven't -> hasn't

File: alf/networks/critic_networks_test.py [hunk 5]
```diff
@@ -118,13 +118,15 @@ def test_make_parallel(self):
 
         def _train(pnet, name):
             t0 = time.time()
-            optimizer = alf.optimizers.AdamTF(list(pnet.parameters()), lr=1e-4)
+            optimizer = alf.optimizers.AdamTF(lr=1e-4)
+            optimizer.add_param_group({'params': list(pnet.parameters())})
```
Comment (Haichao-Zhang) [inline alf/networks/critic_networks_test.py:122]: what's the reason for this change?

Comment (hnyu) [reply alf/networks/critic_networks_test.py:122]: Before the init is wrong because the API is no longer like that. You can still specify params but have to provide the key.

**Review by emailweixu — APPROVED** [review]
